rescan.go

```diff
@@ -277,20 +277,36 @@ func (s *ChainService) Rescan(options ...RescanOption) error {
 		"rescan start (height=%v)", filterHeaderHeight, curStamp.Height)
 
 	// We'll wait here at this point until we have enough filter headers to
-	// actually start walking forwards in the chain.
-	s.blockManager.newFilterHeadersMtx.Lock()
-	for s.blockManager.filterHeaderTip < uint32(curStamp.Height) {
-		s.blockManager.newFilterHeadersSignal.Wait()
+	// actually start walking forwards in the chain. To be able to wake up
+	// in cause we are being asked to exit, we'll launch a new goroutine to
+	// wait.
+	done := make(chan struct{})
+	go func() {
+		s.blockManager.newFilterHeadersMtx.Lock()
+		for s.blockManager.filterHeaderTip < uint32(curStamp.Height) {
+			s.blockManager.newFilterHeadersSignal.Wait()
 
-		// While we're awake, check to see if we need to exit.
-		select {
-		case <-ro.quit:
-			s.blockManager.newFilterHeadersMtx.Unlock()
-			return nil
-		default:
+			// While we're awake, check to see if we need to exit.
+			select {
+			case <-ro.quit:
+				s.blockManager.newFilterHeadersMtx.Unlock()
+				return
+			default:
+			}
 		}
+		s.blockManager.newFilterHeadersMtx.Unlock()
+		close(done)
+	}()
+
+	// Now wait for either filter headers are updated, or we are quitting.
+	select {
+	case <-done:
```

Comment:
@Roasbeef pointed out that we should also be pulling in updates while waiting for the sync to finish. Rn the sync is in the bg, but sends on `updateChan` will still block. Suggested that they can be queued up, and then applied after the sync completes

Reply:
Good idea! Will add